modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProcessor.java

```diff
@@ -28,188 +27,127 @@
+    private final ConcurrentMap<IgniteUuid, ServiceInfo> deployedServices = new ConcurrentHashMap<>();
 
-    /** Busy lock. */
-    private volatile GridSpinBusyLock busyLock = new GridSpinBusyLock();
+    /** Deployment futures. */
+    private final ConcurrentMap<IgniteUuid, GridServiceDeploymentFuture> depFuts = new ConcurrentHashMap<>();
 
-    /** Uncaught exception handler for thread pools. */
-    private final UncaughtExceptionHandler oomeHnd = new OomExceptionHandler(ctx);
+    /** Undeployment futures. */
+    private final ConcurrentMap<IgniteUuid, GridFutureAdapter<?>> undepFuts = new ConcurrentHashMap<>();
 
     /** Thread factory. */
-    private ThreadFactory threadFactory = new IgniteThreadFactory(ctx.igniteInstanceName(), "service",
-        oomeHnd);
+    private final ThreadFactory threadFactory = new IgniteThreadFactory(ctx.igniteInstanceName(), "service",
+        new OomExceptionHandler(ctx));
 
-    /** Thread local for service name. */
-    private ThreadLocal<String> svcName = new ThreadLocal<>();
+    /** Services deployment exchange manager. */
+    private volatile ServicesDeploymentExchangeManager exchMgr = new ServicesDeploymentExchangeManager(ctx);
 
-    /** Service cache. */
-    private volatile IgniteInternalCache<Object, Object> serviceCache;
+    /** Services topologies update mutex. */
+    private final Object servicesTopsUpdateMux = new Object();
 
-    /** Topology listener. */
-    private DiscoveryEventListener topLsnr = new TopologyListener();
+    /** Operations lock. */
+    private final ReentrantReadWriteLock opsLock = new ReentrantReadWriteLock();
 
-    /** */
-    private final CountDownLatch startLatch = new CountDownLatch(1);
+    /** Disconnected flag. */
+    private volatile boolean disconnected;
 
     /**
      * @param ctx Kernal context.
      */
     public GridServiceProcessor(GridKernalContext ctx) {
         super(ctx);
 
-        depExe = Executors.newSingleThreadExecutor(new IgniteThreadFactory(ctx.igniteInstanceName(),
-            "srvc-deploy", oomeHnd));
-
         String servicesCompatibilityMode = getString(IGNITE_SERVICES_COMPATIBILITY_MODE);
 
         srvcCompatibilitySysProp = servicesCompatibilityMode == null ? null : Boolean.valueOf(servicesCompatibilityMode);
     }
 
-    /**
-     * @param ctx Context.
-     * @throws IgniteCheckedException If failed.
-     */
-    public void onContinuousProcessorStarted(GridKernalContext ctx) throws IgniteCheckedException {
-        if (ctx.clientNode()) {
-            assert !ctx.isDaemon();
-
-            ctx.continuous().registerStaticRoutine(
-                CU.UTILITY_CACHE_NAME, new ServiceEntriesListener(), null, null
-            );
-        }
-    }
-
     /** {@inheritDoc} */
     @Override public void start() throws IgniteCheckedException {
         ctx.addNodeAttribute(ATTR_SERVICES_COMPATIBILITY_MODE, srvcCompatibilitySysProp);
 
-        if (ctx.isDaemon())
```

Comment:
Why did you remove this check?

Reply:
`GridServiceProcessor` has been annotated `@SkipDaemon`, so no any additional checks are needed.